### DISCUSSION TITLE
Metadata C++ test is broken

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#183] [Design & Development]
Somewhere in the past someone beside me changed Metadata to change the old overloaded << operator to this:
```
ostringstream& operator<<(ostringstream& os, Metadata& m)
```
That is fine, and I think I may have even changes the test_md.cc code to deal with that, but test_md is most definitely currently broken.   The output of this function is garbled and looks like gibberish.  Here is a typical example:
```
Ym9vbF92YWw. bool 1
ZG91YmxlX3ZhbA.. double 2.5
bG9uZ192YWw. long 10
c3RyaW5nX3ZhbA.. string dGVzdF9zdHJpbmc.
```
which should have been
```
bool_val bool 1
double_val double 2.5
long_val long 10
string_val string test_string
```
The reason is the key is being handled with this line of code in that function:
```
if(pretty_name.find("basic_string")==std::string::npos)
            sname=pretty_name;
os<<misc::base64_encode(mdptr->first.c_str(), mdptr->first.size())<<" "<<sname<<" ";
```
The sname variable is comparable to python's type(a) description - the demangled type name.   That part works and is printed in the example above as "bool", "double", "long", and "string" in sequence. 

Why is the "misc::base64_encode" call in third line of code above?  It is what is clobbering this output in the C++ test program (and when I tried to use it for a test converting by gclgrid library for mspass).   I would just change this function but there must be some reason for that incantation?  Did you have to do that to make this work right on the python side?   Can you please educate me?

**Comment by wangyinz** [comment]
If you check the git blame on that line, you would see why. The `base64_encode` was introduced to resolve a series of issues created by using escape characters. Basically, we are trying to resolve such an issue: 

https://github.com/wangyinz/mspass/blob/3f8924e4dd76da68ca0764582983c52948904291/python/tests/test_ccore.py#L256-L273

If we do not encode those strings, the deserialization will break. Yet, it can still break when the string are empty (i.e., `''`).

As you can see the root of the problem is that we rely on space-delimited strings for serialization, which is a bad practice anyway. So, I recently fixed the problem in #177, which converts the Metadata to dict and then serializes. I did that in a new pair of functions, which means now the old serialization code is not called on the Python side anymore. 

Anyway, long story, but I think the solution here is that you can safely remove everything related to the base64 encoding, and those function should work just fine with C++ code.

**Reply by pavlis** [reply]
Ok, I'm going to fix it then.  I'm also going to fundamentally change the algorithm.   I'm going to just have it loop over the keys and depend on operator overloading for deal with data types.  Well, maybe.  I have to read the documentation again on boost::any.   It has been way too long ago for me to remember how it even worked.

**Reply by pavlis** [reply]
Well, can't do what I was thinking but still can simply the code from earlier - I think.   I could just restore the old version but there are some weird things in there I think worth fixing at the same time.

**Comment by pavlis** [comment]
Well can't close this until we settle another thing that came up immediately when I tried to fix the operator<< C++ function for Metadata.  The fix I did to make that work breaks the methods you added and noted above called "serialize_metadata" and "restore_serialized_metadata".   the reason is that serialize_metadata depends upon the base64_encode approach because it calls operator<<.  Should I:
1.  just remove serialize_metadata and restore_serialized_metadata or
2.  Fix serialize_metadata to utilize the base64_encode stuff?  I think we need to retain that anyway to allow operator<< to work correctly if the value in in a key-value pair is a python object.

I would suggest 1 to preserve this capability in case it came up in a different context - e.g. MPI.
